src/aten/Distributions.cpp

```diff
@@ -0,0 +1,314 @@
+        "), 2^",                                                                                     \
+        digits,                                                                                      \
+        "]. ",                                                                                       \
+        "Due to precision limitations ",                                                             \
+        dtype,                                                                                       \
+        " can support discrete uniform distribution only within this range. ",                       \
+        "This warning will become an error in version 1.7 release, please fix the code in advance"); \
+  }
+
+static void check_from_to_in_range(
+    int64_t from,
+    int64_t to_inc,
+    caffe2::TypeMeta dtype) {
+  const auto scalar_type = typeMetaToScalarType(dtype);
+  if (isFloatingType(scalar_type)) {
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+        at::ScalarType::Half,
+        at::ScalarType::BFloat16,
+        scalar_type,
+        "check_random_fp_bounds",
+        [&] {
+          const auto min =
+              static_cast<double>(std::numeric_limits<scalar_t>::lowest());
+          const auto max =
+              static_cast<double>(std::numeric_limits<scalar_t>::max());
+          CHECK_OUT_OF_BOUNDS(from, "from", min, max, dtype);
+          CHECK_OUT_OF_BOUNDS(to_inc, "to - 1", min, max, dtype);
+
+          constexpr auto digits = std::numeric_limits<scalar_t>::digits;
+          WARN_OUT_OF_BOUNDS(from, "from", digits, dtype);
+          WARN_OUT_OF_BOUNDS(to_inc, "to - 1", digits, dtype);
+        });
+  } else if (scalar_type == kUInt64) {
+    // When you do a comparison between int64_t and uint64_t, the usual
+    // arithmetic conversions say that the int64_t value is promoted to
+    // unsigned. But this conversion wraps around: if I had -1 as my int64_t,
+    // then it will promote to 0xFFFFFFFFFFFFFFFF in uint64_t. This is never
+    // the right thing to do.
+    CHECK_OUT_OF_BOUNDS(from, "from", 0, INT64_MAX, dtype);
+    CHECK_OUT_OF_BOUNDS(to_inc, "to - 1", 0, INT64_MAX, dtype);
+  } else if (isIntegralType(scalar_type, /*includeBool=*/true)) {
+    AT_DISPATCH_V2(
+        scalar_type,
+        "check_random_integral_bounds",
+        AT_WRAP([&]() {
+          const auto min =
+              static_cast<int64_t>(std::numeric_limits<scalar_t>::lowest());
+          const auto max =
+              static_cast<int64_t>(std::numeric_limits<scalar_t>::max());
+          CHECK_OUT_OF_BOUNDS(from, "from", min, max, dtype);
+          CHECK_OUT_OF_BOUNDS(to_inc, "to - 1", min, max, dtype);
+        }),
+        AT_EXPAND(AT_INTEGRAL_TYPES),
+        kUInt16,
+        kUInt32,
+        kBool);
+  } else {
+    TORCH_CHECK(
+        false,
+        "check_random_bounds handles only integral, floating-point and boolean types");
+  }
+}
+
+template <typename scalar_t>
+int64_t update_from(int64_t from) {
```

Comment:
Basic data type arithmetic,
Reuse utils in `ATen/native/DistributionTemplates.h`.